packages/core/src/event_delegation_utils.ts

```diff
@@ -0,0 +1,113 @@
+  if (!handlerFns) {
+    return;
+  }
+  for (const handler of handlerFns) {
+    handler(event);
+  }
+}
+
+export function setJSActionAttribute(nativeElement: Element, eventTypes: string[]) {
+  if (!eventTypes.length) {
+    return;
+  }
+  const parts = eventTypes.reduce((prev, curr) => prev + curr + ':;', '');
+  const existingAttr = nativeElement.getAttribute(Attributes.JSACTION);
+  //  This is required to be a module accessor to appease security tests on setAttribute.
+  nativeElement.setAttribute(Attributes.JSACTION, `${existingAttr ?? ''}${parts}`);
+}
+
+export const sharedStashFunction = (rEl: RElement, eventType: string, listenerFn: () => void) => {
+  const el = rEl as unknown as Element;
+  const eventListenerMap = el.__jsaction_fns ?? new Map();
+  const eventListeners = eventListenerMap.get(eventType) ?? [];
+  eventListeners.push(listenerFn);
+  eventListenerMap.set(eventType, eventListeners);
+  el.__jsaction_fns = eventListenerMap;
+};
+
+export const removeListeners = (el: Element) => {
+  el.removeAttribute(Attributes.JSACTION);
+  el.__jsaction_fns = undefined;
+};
+
+// TODO: Upstream this back into event-dispatch.
+function getJsactionData(container: EarlyJsactionDataContainer) {
+  return container._ejsa;
+}
+
+@Injectable({providedIn: 'root'})
+export class GlobalEventDelegation {
+  eventContract!: EventContract;
+  addEvent(el: Element, eventName: string) {
+    if (this.eventContract) {
+      this.eventContract.addEvent(eventName);
+      setJSActionAttribute(el, [eventName]);
+      return true;
+    }
+    return false;
+  }
+}
+
+export const initEventReplay = (eventDelegation: GlobalEventDelegation, injector: Injector) => {
+  const appId = injector.get(APP_ID);
+  // This is set in packages/platform-server/src/utils.ts
+  // Note: globalThis[CONTRACT_PROPERTY] may be undefined in case Event Replay feature
+  // is enabled, but there are no events configured in an application.
+  const container = globalThis[CONTRACT_PROPERTY]?.[appId];
+  const earlyJsactionData = getJsactionData(container)!;
+  eventDelegation.eventContract = new EventContract(
+    new EventContractContainer(earlyJsactionData.c),
+  );
+  for (const et of earlyJsactionData.et) {
+    eventDelegation.eventContract.addEvent(et);
+  }
+  for (const et of earlyJsactionData.etc) {
+    eventDelegation.eventContract.addEvent(et);
+  }
+  eventDelegation.eventContract.replayEarlyEvents(container);
+  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
+  registerDispatcher(eventDelegation.eventContract, dispatcher);
+};
+
+export const initGlobalEventDelegation = (
+  eventDelegation: GlobalEventDelegation,
+  injector: Injector,
+) => {
+  if (injector.get(IS_EVENT_REPLAY_ENABLED, EVENT_REPLAY_ENABLED_DEFAULT)) {
+    return;
+  }
+  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
+  registerDispatcher(eventDelegation.eventContract, dispatcher);
+};
```

Comment:
I think we can avoid referencing `IS_EVENT_REPLAY_ENABLED` here and instead check if the `eventDelegation.eventContract` is defined:
```suggestion
export const initGlobalEventDelegation = (
  eventDelegation: GlobalEventDelegation,
  eventContractContainer: Element,
) => {
  if (eventDelegation.eventContract) {
    // EventContract is already configured for this application,
    // no need to create a new one, exit.
    return;
  }
  const container = new EventContractContainer(eventContractContainer);
  const eventContract = new EventContract(container);
  const dispatcher = new EventDispatcher(invokeRegisteredListeners);
  registerDispatcher(eventContract, dispatcher);
  eventDelegation.eventContract = eventContract;
};
```
I've also updated the function to init `EventContract` if needed. We can use this function later in the client-only code path. WDYT?

Reply:
i was a little afraid of order of operations in the case both eventReplay() and withGlobalEventDelegation() are available.  I did initialize the event contract though.